File: app/javascript/models/feed_filter.ts
import { FilterableField } from "./filterable_field";

enum FilterPronoun {
  Any = "any_",
  All = "all_",
}

function filterPronounLabel(value: FilterPronoun): string {
  const res = Object.keys(FilterPronoun).find(key => FilterPronoun[key as keyof typeof FilterPronoun] === value);
  if (res === undefined) { throw new Error(`This code should never run, how did you get here with ${value}?`); }
  return res;
}

enum FilterType {
  StartsWith = "Starts with",
  EndsWith = "Ends with",
  Contains = "Contains",
  Matches = "Matches",
}

function filterTypeValue(value: FilterType): string {
  const res = Object.keys(FilterType).find(key => FilterType[key as keyof typeof FilterType] === value);
  if (res === undefined) { throw new Error(`This code should never run, how did you get here with ${value}?`); }
  return res;
}

type FilterSubstitution = {
  field: FilterableField;
  value: string;
};

type FilterCondition = {
  filter_type: FilterType;
  field: FilterableField;
  value: string;
};

type FeedFilter = {
  id: number;
  url: string;
  pronoun: FilterPronoun,
  conditions: FilterCondition[],
  substitutions: FilterSubstitution[],
  feed_id: number;
};

type FeedFilterForCreation = {
  id?: number;
  url: string;
  pronoun: FilterPronoun,
  conditions: FilterCondition[],
  substitutions: FilterSubstitution[],
  feed_id?: number;
};

export {
  FilterPronoun,
  filterPronounLabel,
  FilterType,
  filterTypeValue,
  FilterSubstitution,
  FilterCondition,
  FeedFilter,
  FeedFilterForCreation,
};

File: app/models/feed_filter.rb
class FeedFilter < ApplicationRecord
  belongs_to :feed
  after_initialize :init

  validates :url, presence: true
  validates :pronoun, presence: true
  validates :feed_id, presence: true
  validate :conditions_presence
  validate :substitutions_presence

  # Stupid activerecord already has a method named #all, so I can't name my enum correctly...
  enum pronoun: {
    any_: 0,
    all_: 1,
    none_: 2,
  }

  private

  def init
    self.url ||= ""
    self.pronoun ||= FeedFilter.pronouns.keys.first
    self.conditions ||= [FilterCondition.default_hash]
    self.substitutions ||= [] # [FilterSubstitution.default_hash]
  end

  def conditions_presence
    errors.add(:conditions, "must have at least one entry") if conditions.blank?
    blank_errors = []
    conditions.each do |condition|
      blank_errors << "'#{condition["field"]} #{condition["filter_type"].downcase}' must have a value" if condition["value"].blank?
    end
    blank_errors.to_set.each do |error|
      errors.add(:conditions, error)
    end
  end

  def substitutions_presence
    blank_errors = []
    substitutions.each do |substitution|
      blank_errors << "'#{substitution["field"]}' must have a value" if substitution["value"].blank?
    end
    blank_errors.to_set.each do |error|
      errors.add(:substitutions, error)
    end
  end
end
